hardware/raspberrypi/bootmodes/nvme.md

```diff
@@ -0,0 +1,154 @@
+# NVMe SSD Boot
+
+NVMe (Non Volatile Memory express) is a standard for accessing Solid State Drives (SSD) via a PCI express bus (PCIe). You can connect these drives via the exposed PCIe slot on a Compute Module (CM) IO board allowing a CM4 to boot from SSD.
+
+## Hardware Needed
+
+You need an NVMe M.2 SSD. You can’t plug an M.2 SSD directly into the PCIe slot on the IO board so you’ll need an adapter. Be careful to get the right one. Search for PCI-E 3.0 x1 Lane to M.2 NGFF M-Key SSD Nvme PCI Express Adapter Card.
+
+You don't need it, but if you want to access an NVMe drive from a PC, you might want to get a USB adapter. Search for NVME PCI-E M-Key Solid State Drive External Enclosure.
+
+Linux supports NVMe drives so you should be able to plug it in to the IO board and see your NVMe device to check it works.
+
+```
+$ ls -l /dev/nvme*
+crw------- 1 root root 245, 0 Mar  9 14:58 /dev/nvme0
+brw-rw---- 1 root disk 259, 0 Mar  9 14:58 /dev/nvme0n1
+```
+
+## Software Needed
+
+To boot from NVMe you need pre-release versions of the bootloader, VideoCore firmware and kernel.
+
+### Bootloader
+
+You need to use [usbboot](https://github.com/raspberrypi/usbboot) to update the CM4 bootloader. See the instructions for how to build rpiboot and configure the IO board to switch the ROM to usbboot mode. You can then update your bootloader to support nvme boot with the following command.
+
+```
+$ ./rpiboot -d nvme/
+
+Loading: nvme//bootcode4.bin
+Waiting for BCM2835/6/7/2711...
+Loading: nvme//bootcode4.bin
+Sending bootcode.bin
+Successful read 4 bytes
+Waiting for BCM2835/6/7/2711...
+Loading: nvme//bootcode4.bin
+Second stage boot server
+Loading: nvme//config.txt
+File read: config.txt
+Loading: nvme//pieeprom.bin
+Loading: nvme//pieeprom.bin
+Loading: nvme//pieeprom.sig
+File read: pieeprom.sig
+Loading: nvme//pieeprom.bin
+File read: pieeprom.bin
+Second stage boot server done
+```
+
+This command also updates the boot order so it will try to boot from NVMe AFTER the SD card but before USB and the network.
+
+### Firmware and Kernel
+
+At the time of writing you need to use pre-release versions of the VideoCore firmware and kernel and these need to be on the NVMe disk. When this software is officially released your will just be able to write the software to the NVMe disk directly with the rpi-imager app. Until then the easiest way to do this is to..
+
+* Write Raspberry PI OS to a spare SD card with the [rpi-imager app](https://github.com/raspberrypi/rpi-imager) and boot the board with this.
+* Update the software on the SD card with the latest pre-release software using [rpi-update](../../../raspbian/applications/rpi-update.md). Pre-release software might be buggy so don't use this on any image you can't afford to lose!
+*  Then use [piclone](https://github.com/raspberrypi-ui/piclone) to copy the image to the NVMe disk. When using piclone choose the option to use "New Partition ids"
```

Comment:
Please put things like "piclone" in backticks `piclone` - that highlights them nicely as commands. Not needed in the link here, just the reference.

Reply:
Thanks - I've used this in a few places

Reply:
Thanks - I've made us of this